Refactor IMAP commands

rhari991 opened the pull request:
This PR creates separate classes for the UID SEARCH, UID FETCH, UID STORE and UID COPY commands that take the command's parameters as instance variables and create the String representation of the command. They also handle splitting the command if it becomes too long.

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/command/BaseCommand.java:
```diff
@@ -0,0 +1,55 @@
+
+        List<BaseCommand> commands;
+        String commandString = createCommandString();
+        if (commandString.getBytes().length > getCommandLengthLimit()) {
```
how about `String.length()`? :)

rhari991 replied:
Done

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/command/UidSearchCommand.java:
```diff
@@ -0,0 +1,409 @@
+    private void addRequiredFlags(StringBuilder builder) {
+        if (requiredFlags != null) {
+            for (Flag flag : requiredFlags) {
+                switch (flag) {
```
this switch statement deserves its own method

rhari991 replied:
I was thinking of moving the strings into instance variables in the Flag enum as well

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/command/UidSearchCommand.java:
```diff
@@ -0,0 +1,409 @@
+    private void addForbiddenFlags(StringBuilder builder) {
+        if (forbiddenFlags != null) {
+            for (Flag flag : forbiddenFlags) {
+                switch (flag) {
```
same here

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/command/UidSearchCommand.java:
```diff
@@ -0,0 +1,409 @@
+    private MessageRetrievalListener<ImapMessage> listener;
+
+    private Set<Long> sequenceSet;
+    private List<Map.Entry<String, String>> sequenceRanges;
```
hmh. I see how this is more statically typed, but the readability does suffer compared to a simpler `List<String>` with a `.size() % 2 == 0` invariant to the builder

rhari991 replied:
I was thinking of making this into `List<Map.Entry<Long, Long>>`  to make it easier for comparison while minimizing the number of elements.

rhari991 replied:
Made a custom class for this.

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/command/UidSearchCommand.java:
```diff
@@ -0,0 +1,409 @@
+        addSince(builder);
+        addRequiredFlags(builder);
+        addForbiddenFlags(builder);
+        if (builder.lastIndexOf(" ") == builder.length() - 1) {
```
`Builder.chatAt() == ' '`

rhari991 replied:
Replaced with trim() instead.

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/command/UidSearchCommand.java:
```diff
@@ -0,0 +1,409 @@
+                        .uidRanges(null);
+
+                int length = builder.getLength();
+                while (length < lengthLimit) {
```
so this increases the length until it's above the length limit? I'm not sure if that's a problem here, but if that's the intent I guess "threshold" is a better name :)

rhari991 replied:
Was an oversight, fixed so it stops before the limit.

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/command/UidSearchCommand.java:
```diff
@@ -0,0 +1,409 @@
+        }
+
+        private int getLength() {
+            return this.build().createCommandString().getBytes().length;
```
hmm, this method seems like a hack. looking at its use above, maybe change the loop to a `do .. while();` that builds at the end, checks the length and then uses the result that was built

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/response/BaseResponse.java:
```diff
@@ -0,0 +1,91 @@
+        return commandFactory.getLogId();
+    }
+
+    void extractExtras(List<ImapResponse> responses) {
```
why isn't this called handleUntaggedResponses? seems to be exactly what it does

rhari991 replied:
Done

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/response/SearchResponse.java:
```diff
@@ -0,0 +1,71 @@
+
+        List<Long> numbers = new ArrayList<>();
+
+        //Currently searching is done only on the basis of either uids or sequence numbers
```
perfect comment granularity :+1:

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/response/SearchResponse.java:
```diff
@@ -0,0 +1,71 @@
+    }
+
+    private static void parseSingleLine(ImapResponse response, List<Long> numbers) {
+        if (response.isTagged() || response.size() < 2 || !equalsIgnoreCase(response.get(0), Responses.SEARCH)) {
```
Are there circumstances where this happens and it's not the imap server misbehaving? I feel like this should be a `Timber.w` or even an exception, if we are really getting bogus responses.

(I know this line is not yours, just bringing it up since we're touching it)

Valodim commented:
I'd prefer Google's [AutoValue](https://github.com/google/auto/tree/master/value) to lombok, because it's from google, we use it in openkeychain, and the sparse documentation of lombok was an immediate turn-off.

rhari991 commented:
@Valodim @philipwhiuk @cketti Should be good for a review !

rhari991 commented on k9mail-library/src/test/java/com/fsck/k9/mail/store/imap/ImapFolderTest.java:
```diff
@@ -478,7 +478,7 @@ public void getUnreadMessageCount() throws Exception {
         ImapFolder folder = createFolder("Folder");
```
Some tests, like this one, are behaving flakily. Can anyone help me with this ?

Valodim commented:
Alright! Will do! :)

rhari991 commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/SelectedStateCommand.java:
```diff
@@ -0,0 +1,296 @@
+    }
+
+    public SelectedStateResponse execute() throws MessagingException {
+        return null;
```
It's meant to be a method that does not necessarily need to overridden. I saw this pattern somewhere in the k-9 code. The search and copy commands return parsed responses and so override this. The fetch command is never executed as it directly uses the connection object. The store method needs to be executed but the response is not used.

rhari991 commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/SelectedStateCommand.java:
```diff
@@ -0,0 +1,296 @@
+    calculating the length of the command)
+     */
+    private static final int LENGTH_LIMIT_WITHOUT_CONDSTORE = 980;
+    private static final int LENGTH_LIMIT_WITH_CONDSTORE = 8172;
```
Not sure if it is worth mentioning as it is not relevant to the code and is present in the RFC. It is mostly because of the QRESYNC extension, which modifies the SELECT/EXAMINE commands to provide all known message sequence numbers and their corresponding UIDs as arguments.

philipwhiuk replied:
I disagree. Also I think it's worth mentioning the point in the RFC where the limit is defined.

rhari991 commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/SelectedStateCommand.java:
```diff
@@ -0,0 +1,296 @@
+            }
+
+        } else {
+            //This should never happen
```
Will do

rhari991 commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/SelectedStateCommand.java:
```diff
@@ -0,0 +1,296 @@
+
+    }
+
+    private static class Range {
```
Changed to `ContiguousIdGroup`

rhari991 commented:
- Made a separate `ImapCommandSplitter` class.
- I've removed a lot of the code in those classes.
- Nice catch, I did not see that the folder and connection objects could be passed in the execute method. Modified that as well.

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/ImapCommandSplitter.java:
```diff
@@ -0,0 +1,106 @@
+
+class ImapCommandSplitter {
+
+    static List<SelectedStateCommand> splitCommand(SelectedStateCommand command, int lengthLimit) {
```
where's the tests? :)

rhari991 replied:
I've tested this through the SelectedStateCommand tests. Should I move those ?

Valodim replied:
Yes. Nicely encapsulated functionality like this should definitely be tested individually. It's also nice being able to jump to the tests for a class with `Ctrl+Shift+T` if the name matches

rhari991 replied:
@Valodim done !

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/UidFetchCommand.java:
```diff
@@ -0,0 +1,147 @@
+import com.fsck.k9.mail.store.imap.ImapUtility;
+
+
+public class UidFetchCommand extends SelectedStateCommand {
```
no tests?

rhari991 replied:
Won't the current tests in `ImapFolderTest` cover this ? They either mock the `imapConnection.executeSimpleCommand` method to return mock responses when the correct command string is provided as an argument, or directly check if it was invoked with the correct command.

Valodim replied:
You're right, that'll do :+1:

philipwhiuk replied:
I'd rather see per class tests and make ImapFolderTest do less.

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/UidSearchCommand.java:
```diff
@@ -0,0 +1,191 @@
+import com.fsck.k9.mail.store.imap.selectedstate.response.UidSearchResponse;
+
+
+public class UidSearchCommand extends SelectedStateCommand {
```
no tests?

Valodim replied:
note that those don't have to be fancy: just call the setters and see that it returns the expected string. that already covers a lot of ground

philipwhiuk commented on k9mail-library/src/test/java/com/fsck/k9/mail/store/imap/ImapResponseHelper.java:
```diff
@@ -3,16 +3,37 @@
 public class ImapResponseHelper {
+
+    @SafeVarargs
+    public static List<List<ImapResponse>> createMultipleImapResponses(List<String>... responses) throws IOException {
```
Why is it a list of lists, not just a list?

rhari991 replied:
The selected state commands split the commands and execute them individually, so you get a `List<List<ImapResponse>>`. These are then parsed by the response classes and combined into a single response object which is returned.

philipwhiuk commented on k9mail-library/src/test/java/com/fsck/k9/mail/store/imap/ImapResponseHelper.java:
```diff
@@ -3,16 +3,37 @@
+        return imapResponses;
+    }
+
+    public static List<ImapResponse> createImapResponse(List<String> responses) throws IOException {
```
This is misleading given it returns multiple ImapResponses

rhari991 replied:
changed

philipwhiuk commented on k9mail-library/src/test/java/com/fsck/k9/mail/store/imap/selectedstate/response/UidSearchResponseTest.java:
```diff
@@ -0,0 +1,136 @@
+
+    @Test
+    public void parse_withMultipleSearchResponses_shouldExtractNumbers() throws Exception {
+        List<List<ImapResponse>> responses = createMultipleImapResponses(SEARCH_RESPONSE_1, SEARCH_RESPONSE_2, SEARCH_RESPONSE_3);
```
This and other lines are too long (if there's a scrollbar in GitHub the line is too long).

rhari991 replied:
fixed

philipwhiuk commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/SelectedStateCommand.java:
```diff
@@ -0,0 +1,204 @@
+import com.fsck.k9.mail.store.imap.selectedstate.response.SelectedStateResponse;
+
+
+//This is the base class for a command that is used in the "selected state" i.e when a mailbox is selected
```
FolderSelectedStateCommand and remove the comment.

rhari991 replied:
done

philipwhiuk commented on k9mail-library/src/test/java/com/fsck/k9/mail/store/imap/ImapFolderTest.java:
```diff
@@ -462,7 +462,7 @@ public void getUnreadMessageCount_withClosedFolder_shouldThrow() throws Exceptio
     public void getUnreadMessageCount_connectionThrowsIOException_shouldThrowMessagingException() throws Exception {
         ImapFolder folder = createFolder("Folder");
         prepareImapFolderForOpen(OPEN_MODE_RW);
-        when(imapConnection.executeSimpleCommand("SEARCH 1:* UNSEEN NOT DELETED")).thenThrow(new IOException());
```
I'd rather see us add tests than change ones if it's new behaviour. Do we never use non-UID based search now?

rhari991 replied:
No, we do not.

rhari991 replied:
@philipwhiuk The tests I've added should cover everything now.

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/UidCopyCommand.java:
```diff
@@ -0,0 +1,60 @@
+        return new Builder().destinationFolderName(destinationFolderName);
+    }
+
+    public static class Builder extends FolderSelectedStateCommand.Builder<UidCopyCommand, Builder> {
```
shouldn't a factory method work just fine here?

rhari991 replied:
@Valodim do you mean making two methods

`createWithUids(Set<Long> uids, String destinationFolderName)' 
and
`createWithUidRange(Long start, Long end, String destinationFolderName)'   ?

Valodim replied:
something like that, yes

rhari991 replied:
The `newBuilder()` method above it is necessary as it is a concrete class. So it's not possible to directly replace the Builder.

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/FolderSelectedStateCommand.java:
```diff
@@ -0,0 +1,207 @@
+
+        @Override
+        public String toString() {
+            if (start == LAST_ID && end == LAST_ID) {
```
shouldn't this be `start == FIRST_ID`?

rhari991 replied:
Nope, this is meant to cover the method where UID SEARCH is used to find the highest UID. The case 1:* is covered by the block below this.

Valodim replied:
Oh, so `*:*` is the highest? I would have assumed it returns all messages

rhari991 replied:
Yeah I thought that at first as well. The order of the arguments when specifying an id range is not important, so 5:10 is the same as 10:5 . So *:* locks you to the highest id.

Valodim replied:
that's fine, then. comment would be helpful to make it clear :)

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/ImapCommandSplitter.java:
```diff
@@ -0,0 +1,106 @@
+        return commands;
+    }
+
+    private static FolderSelectedStateCommand optimizeGroupings(FolderSelectedStateCommand command) {
```
this method should have a test or two of its own

rhari991 replied:
added

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/ImapCommandSplitter.java:
```diff
@@ -0,0 +1,106 @@
+    static List<FolderSelectedStateCommand> splitCommand(FolderSelectedStateCommand command, int lengthLimit) {
+        List<FolderSelectedStateCommand> commands = new ArrayList<>();
+
+        if (command.getIdSet() != null || command.getIdGroups() != null) {
```
if you invert this `if` and throw, you can save one indentation level for the rest of the method

rhari991 replied:
done

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/UidSearchCommand.java:
```diff
@@ -0,0 +1,191 @@
+            return builder;
+        }
+
+        public Builder queryString(String queryString) {
```
so much boilerplate :\ would this be easier to do with [AutoValue](https://github.com/google/auto/blob/master/value/userguide/index.md)?

rhari991 replied:
Does not seem to be worth it.

Valodim commented:
@philipwhiuk were all points in your review addressed?

rhari991 commented:
@Valodim @philipwhiuk Please review this again (last two commits). I have rewritten most of this to not use the Builder pattern.

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/FolderSelectedStateCommand.java:
```diff
@@ -0,0 +1,172 @@
+    }
+
+    public SelectedStateResponse execute(ImapConnection connection, ImapFolder folder) throws MessagingException {
+        return null;
```
should this ever happen? if not, just leave out the method? this is an abstract class, after all

Valodim commented on k9mail-library/src/main/java/com/fsck/k9/mail/store/imap/selectedstate/command/FolderSelectedStateCommand.java:
```diff
@@ -0,0 +1,172 @@
+        return null;
+    }
+
+    List<List<ImapResponse>> executeInternal(ImapConnection connection, ImapFolder folder)
```
I don't quite like that the responsibility for executing the actual call on the `ImapConnection` lies in this object. I'd rather this be part of an `ImapConnection.execute` method.

As a suggestion:
```
abstract class <R extends FolderSelectedResponse> FolderStateSelectedCommand<R> {
  abstract R parseResponse(List<ImapResponse> unparsedResponses);
}

public class ImapConnection {
  public <R extends FolderSelectedResponse> R execute(FolderSelectedStateCommand<R> command) {
    // roughly what executeInternal does
    // obtains the list of commands from command.getCommands
    // parses the response from command.parseResponse to get the result
  }
}
```

That way, the command classes only provide the commands as strings, and a way to parse the responses into specialized objects. Thoughts?

rhari991 replied:
@Valodim I'm not sure if the parsing logic should be in the Command or the Response classes, both seem fine to me. I'm leaning slightly towards the response classes though.

cketti commented:
I'd like to abandon this in favor of PR #2924.

Valodim commented:
:+1:
